### PR TITLE
External Dynamic Manila Provisioner

### PR DESCRIPTION
**Pull request by pospispa**
Trello card: https://trello.com/c/2GAHIEmX/350-13-implement-dynamic-provisioner-for-manila-external-provisioner

### CODE REVIEW

**Comment by pospispa** [comment]
@jsafrane @wongma7 @rootfs PTAL

**Comment by wongma7** [comment]
install https://github.com/sgotti/glide-vc and run `glide-vc --use-lock-file` to remove the extra changes to vendor dir

**Comment by pospispa** [comment]
@wongma7 thank you for the hint `glide-vc --use-lock-file`, however, `glide-vc` removes packages that are needed even though they are in both `glide.lock` and `glide.yaml` files. For example `github.com/go-openapi/spec`:
```
Running unit tests in ./lib/controller
# github.com/kubernetes-incubator/external-storage/lib/controller
package github.com/kubernetes-incubator/external-storage/lib/controller (test)
	imports k8s.io/client-go/kubernetes/fake
	imports k8s.io/api/admissionregistration/v1alpha1
	imports k8s.io/apimachinery/pkg/apis/meta/v1
	imports github.com/go-openapi/spec: cannot find package "github.com/go-openapi/spec" in any of:
	/home/ppospisi/src/k8s.io/apimachinery/vendor/github.com/go-openapi/spec (vendor tree)
	/usr/local/go/src/github.com/go-openapi/spec (from $GOROOT)
	/home/ppospisi/src/github.com/go-openapi/spec (from $GOPATH)
FAIL	github.com/kubernetes-incubator/external-storage/lib/controller [setup failed]
```

That's why I'll rebase my branch without using the `glide-vc --use-lock-file` command.

File: glide.yaml [hunk 1]
```diff
@@ -10,24 +10,38 @@ import:
...
+  version: e509bb64fe1116e12a32273a2032426aa1a5fd26
 - package: k8s.io/apimachinery
-  version: kubernetes-1.8.2
+  version: 019ae5ada31de202164b118aee88ee2d14075c31
 - package: k8s.io/client-go
-  version: kubernetes-1.8.2
+  version: 35ccd4336052e7d73018b1382413534936f34eee
 - package: k8s.io/api
-  version: kubernetes-1.8.2
+  version: 4df58c811fe2e65feb879227b2b245e4dc26e7ad
```
Comment (wongma7) [inline glide.yaml]: use a release tag instead? otherwise it's hard to tell whether the repos are in sync

**Comment by wongma7** [comment]
OK, I will look into why glide-vc doesn't work here

**Comment by jsafrane** [comment]
Would it be possible to make the PR smaller? Perhaps by updating existing dependencies in a separate PR and have hopefully much  smaller PR with the code + new dependencies? Github can't handle this PR:
`The diff you're trying to view is too large. We only load the first 3000 changed files.`

**Comment by childsb** [comment]
Same feature here: https://github.com/kubernetes-incubator/external-storage/pull/429

I asked in #429 to merge the efforts into one..  @chestack, @wongma7 and @pospispa are you guys available for a call Monday morning to discuss?

**Comment by pospispa** [comment]
@childsb I assume it's difficult to meet for a call as you and @wongma7 are in CST (UTC-6) time zone, I'm in CET (UTC+1) time zone and I assume @chestack is in China time zone (UTC+8).

Anyway, I'll be available on Monday from 10:30 till 17:00 CET.

**Comment by pospispa** [comment]
The Manila external dynamic provisioner reads authentication data from environment variables. There is a variable for project/tenant name. So the Manila provisioner can provision shares in 1 project only. However, typical Manila service installation has 10th to 100th projects.

IMHO, as the proposed Manila provisioner is an engineering preview the unability to provisiong shares in more than 1 project is currently not an issue, however, will be definitely an issue for production.

I'd like to hear opinions from reviewers.

**Comment by jsafrane** [comment]
regarding #429 and #524, both seem to implement the same thing, #524 has more unit tests and better error checking all over the place, however the code style is better in #429.

@chestack, can you please look at this PR if there is something obvious missing here?

**Comment by chestack** [comment]
Sure, I am reviewing this large PR.

File: openstack-sharedfilesystems/cmd/manila-provisioner/main.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+	if pvc.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+
+	client := createManilaV2Client()
+
+	var err error
+	var createdShare shares.Share
+	var createReq shares.CreateOpts
+	if createReq, err = sharedfilesystems.PrepareCreateRequest(pvc, devMockGetAllZones); err != nil {
```
[the remark this answers is not in the document]
Comment (pospispa) [reply openstack-sharedfilesystems/cmd/manila-provisioner/main.go]: The `createReq` variable is used outside of the scope of the if that's why `:=` is not used.

**Comment by pospispa** [comment]
@jsafrane I believe I've addressed all your comments. Should I squash the commits?

**Comment by tsmetana** [comment]
I tested the provisioner with OpenStack (Pike) and the basic functionality is working well.
/lgtm

**Comment by pospispa** [comment]
@jsafrane plese, would you also approve this PR as @tsmetana or do you have any further comments?

**Pull request merged**